### ISSUE TITLE
Add unsafe module for useful but unlawful optics

### ISSUE THREAD

**Issue by julien-truffaut** [julien-truffaut/Monocle#379]
Candidates are:
- `def unsafeSelect[A](predicate: A => Boolean): Prism[A, A]`
- `def unsafeHCompose[S, A](xs: Lens[S, A]*): Traversal[S, A]` horizontal `Lens` composition

Since each of these optics is potentially unlawful, we will need to provide a good documentation to explain why and when they are safe to use, e.g.

```scala
case class Person(firstName: String, age: Int)
// legal
(unsafeSelect(_.age >= 18) composeLens GenLens[Person](_.name)).modify("adult " + _)

// illegal
(unsafeSelect(_.age >= 18) composeLens GenLens[Person](_.age)).set(0)
```

**Comment by jule64** [comment]
Hi Julien,  I am working on the doc for the UnsafeHCompose and wanted to check with you that my understanding of unsafe use of the method is correct using the example I am working on for the doc:

```
case class SodaCan(name: String, calories: Int, sugars: Int)

val nameL: Lens[SodaCan, String] = GenLens[SodaCan](_.name)
val caloriesL: Lens[SodaCan, Int] = GenLens[SodaCan](_.calories)
val sugarL: Lens[SodaCan, Int] = GenLens[SodaCan](_.sugars)

val coke = SodaCan("CocaCo",657,33)

UnsafeHCompose.unsafeHCompose(caloriesL, caloriesL).modify(_ - 200)(coke)

```
The last line returns `SodaCan = SodaCan(CocaCo,457,33)`  is this an unasafe usage?
Thanks

**Comment by jule64** [comment]
Also since I am using just two lenses I wanted to show how to do it with Traversal instead but struggling to get the right statement i think:
```
val t = Traversal.apply2[SodaCan,Int](sc => sc.calories,sc => sc.calories)((c1,c2,sc)=>SodaCan("Coke",c1,sc.sugars))
t.modify(_-200)(coke)
```

**Comment by julien-truffaut** [comment]
You're correct, `unsafeHCompose(caloriesL, caloriesL)` is illegal because it breaks `modifyGetAll` [law](https://github.com/julien-truffaut/Monocle/blob/master/core/shared/src/main/scala/monocle/law/TraversalLaws.scala#L16-L17). However, `unsafeHCompose(caloriesL, sugarL)` is completely fine since all lenses have different targets.

You could also write `unsafeHCompose(caloriesL, sugarL)` using `Traversal.apply2` e.g.
```scala
Traversal.apply2[SodaCan,Int](sc => sc.calories,sc => sc.sugar)((c,s,sc)=>sc.copy(calories = c, sugar = s))
```

**Comment by jule64** [comment]
I tried running the unsafe example against that law but I am not getting errors or warnings, what should I expect to see? -
```
TraversalLaws(UnsafeHCompose.unsafeHCompose(caloriesL, caloriesL)).modifyGetAll(coke,c => c-200)
res0: monocle.internal.IsEq[List[Int]] = IsEq(List(457, 457),List(457, 457))
```

**Comment by jule64** [comment]
Also adding the implementation we worked on since I haven't committed it yet:
```
object UnsafeHCompose {
  def unsafeHCompose[S, A](xs: Lens[S, A]*): Traversal[S, A] = {
    val lenses = xs.toList
    new PTraversal[S, S, A, A] {
      def modifyF[F[_]: Applicative](f: A => F[A])(s: S): F[S] = {
        val as: List[(F[A], Lens[S, A])] = lenses.map(lens => (f(lens.get(s)), lens))
        as.foldLeft(Applicative[F].pure(s)){ case (fs, (fa, lens)) =>
          Applicative[F].apply2(fa, fs)((a, s) => lens.set(a)(s))
        }
      }
    }
  }
}
```

**Comment by julien-truffaut** [comment]
It is very interesting, I tried to break `modifyGetAll` and `composeModify` laws but they seem to hold. I expected them to fail because `Setter` do not compose horizontally see https://github.com/julien-truffaut/Monocle/issues/213. However, in this case we do not compose optics using `modify` but `modifyF` which means we cannot `get` from a previously modified `S`. 

I will try to find more about this problem, maybe in haskell-lens irc channel.

**Comment by jule64** [comment]
Cool, keen to know what you find!

**Comment by julien-truffaut** [comment]
sorry for the delay @jule64, I talked with edward the author of haskell lens. It is indeed correct that  composing lenses horizontally to get a traversal lead to an illegal traversal if two lenses point to the same data. The counter example rely on the use of `modifyF` with some fancy `Applicative` effect, e.g. the one use to make partsOf #384.

I am tempted to say that the danger of such a `Traversal` is not worth the trouble of defining it in an `unsafe` module. So if everyone is fine with it, I would rather define `unsafeHCompose` in the `Traversal` companion object and maybe we could name it `applyN` and mention in a comment that it is only a legal `Traversal` if all lenses do not overlap.

**Comment by jule64** [comment]
Thanks for the update.  I would be happy to add the applyN into the main module.  Also since we don't break the Traversal laws do we still need to create tests for applyN?

**Comment by julien-truffaut** [comment]
I think you could add a test similar to https://github.com/julien-truffaut/Monocle/blob/master/test/shared/src/test/scala/monocle/TraversalSpec.scala#L29

**Comment by jule64** [comment]
Given we want to add `applyN` in core, does it make sense for `applyN` to be defined in `PTraversal` or only `Traversal` ?
```
def applyN[S, T, A, B](xs: PLens[S,T,A,B]*): PTraversal[S, T, A, B]

def applyN[S, A](xs: Lens[S, A]*): Traversal[S, A]
```

We know the bottom one works but I am struggling with an implementation of the top one, so far I have something like this:
```
  def applyPN[S, T, A, B](xs: PLens[S,T,A,B]*): PTraversal[S, T, A, B] = {
    val lenses = xs.toList
    new PTraversal[S, T, A, B] {
      def modifyF[F[_]: Applicative](f: A => F[B])(s: S): F[T] = {
        val as: List[(F[B], PLens[S,T, A,B])] = lenses.map(lens => (f(lens.get(s)), lens))
        as.foldLeft(Applicative[F].pure(s)){ case (fs, (fa, lens)) =>
          Applicative[F].apply2(fa, fs)((a, s) => lens.set(a)(s))
        }
      }
    }
  }
```
But `lens.set(a)(s))` wants a `B` not an `A`..

**Comment by julien-truffaut** [comment]
I don't think it is possible to implement `applyN` for `PTraversal` because we need to call `set` n times one after the other. However with `PTraversal` once we call `set` we get a `T` which block us from calling `set` again.

**Comment by julien-truffaut** [comment]
@jule64 do you need some help to finish the PR? I think you were pretty close. If you don't have time, would you mind to just send a PR with what you have and I can polish it later. Thanks again!

**Comment by jule64** [comment]
Thanks for checking, I think I was a bit stuck but I will give it another go and try to send a PR shortly.

**Comment by julien-truffaut** [comment]
fixed by #394
